File: find_keyword_in_commit.py
import git
import os
import re

def filter_commits(search_string):
    # 使用当前目录作为仓库路径
    repo_path = os.getcwd()

    # 检查当前目录是否是 Git 仓库
    if not os.path.exists(os.path.join(repo_path, '.git')):
        print(f"错误: 当前目录 {repo_path} 不是一个 Git 仓库")
        return

    # 初始化仓库
    repo = git.Repo(repo_path)

    # 获取所有提交
    commits = list(repo.iter_commits('--all'))

    # 创建不区分大小写的正则表达式模式
    pattern = re.compile(search_string, re.IGNORECASE)

    # 筛选提交
    filtered_commits = []
    for commit in commits:
        if (pattern.search(commit.author.name) or
            pattern.search(commit.author.email) or
            pattern.search(commit.message)):
            filtered_commits.append(commit)

    # 打印筛选结果
    if filtered_commits:
        print(f"找到 {len(filtered_commits)} 个包含 '{search_string}' 的提交 (不区分大小写):")
        for commit in filtered_commits:
            print(f"提交: {commit.hexsha}")
            print(f"作者: {commit.author.name} <{commit.author.email}>")
            print(f"日期: {commit.authored_datetime}")
            print(f"消息: {commit.message.strip()}")
            print("-" * 40)
    else:
        print(f"没有找到包含 '{search_string}' 的提交 (不区分大小写)")

if __name__ == "__main__":
    search_string = input("请输入要搜索的字符串: ")
    filter_commits(search_string)

File: find_keyword_in_commit.ts
import { $ } from "@david/dax";
import { resolve } from "@std/path";
import { Command } from "@cliffy/command";
import { colors } from "@cliffy/ansi/colors";

const filterCommits = async (pattern: string, dir: string) => {
  try {
    await $`git -C ${dir} rev-parse --is-inside-work-tree`.quiet();

    const gitOutput =
      await $`git -C ${dir} log --all --format="%H%n%an%n%ae%n%ad%n%s%n--COMMIT--"`
        .text();
    const commits = gitOutput
      .split("--COMMIT--")
      .filter((commit) => commit.trim())
      .map((commit) => {
        const [hash, name, email, date, message] = commit.trim().split("\n");
        return { hash, name, email, date, message };
      });

    const re = new RegExp(pattern, "i");
    const filteredCommits = commits.filter(
      (commit) => re.test(commit.message),
    );

    if (filteredCommits.length > 0) {
      console.log(`Found ${filteredCommits.length} commits:`);
      console.log("-".repeat(48));
      for (const commit of filteredCommits) {
        console.log(`Commit: ${commit.hash}`);
        console.log(`Author: ${commit.name} <${commit.email}>`);
        console.log(`Date: ${commit.date}`);
        const highlightedMessage = commit.message.replace(
          re,
          (match) => colors.bold.brightGreen(match),
        );
        console.log(`Message: ${highlightedMessage}`);
        console.log("-".repeat(48));
      }
    } else {
      console.log("No commits found");
    }
  } catch (error) {
    console.error("An error occurred:\n", error);
  }
};

const run = async (_options: unknown, pattern: string, dir = ".") => {
  try {
    await filterCommits(pattern, resolve(dir));
  } catch (error) {
    console.error("An error occurred:", error);
  }
};

if (import.meta.main) {
  await new Command()
    .name("find_keyword_in_commit")
    .description("Find a keyword in commit messages.")
    .arguments("<pattern> [dir]")
    .action(run)
    .parse(Deno.args);
}
